File: 2-Regressão Logística/logist_regression.py
import numpy as np

class LogisticReg:
    """
    Uma implementação simples de Regressão Logística para classificação binária.
    
    A regressão logística é um modelo estatístico usado para prever a probabilidade de um 
    evento de dois resultados (classe 0 ou classe 1) com base em variáveis independentes.

    A classe LogisticReg implementa o algoritmo de otimização de gradiente descendente para
    aprender os parâmetros do modelo (theta) a partir dos dados de treinamento.
    """
    
    def __init__(self, learning_rate=0.01, iterations=1000):
        """
        Inicializa os parâmetros do modelo de Regressão Logística.
        
        Parâmetros:
        learning_rate (float): Taxa de aprendizado do gradiente descendente. Controla o tamanho 
                               do passo durante a atualização dos parâmetros.
        iterations (int): Número de iterações (ou épocas) do processo de otimização.
        """
        self.intercept = None  # Interceptação (termo bias) do modelo
        self.theta = None       # Coeficientes (pesos) do modelo
        self.learning_rate = learning_rate  # Taxa de aprendizado
        self.iterations = iterations  # Número de iterações

    def sigmoid(self, z):
        """
        Função sigmoide que transforma um valor real em um valor entre 0 e 1.
        
        A função sigmoide é usada para modelar a probabilidade de um evento ocorrer.
        
        Parâmetros:
        z (ndarray): Vetor ou valor escalar para o qual a função sigmoide será aplicada.

        Retorna:
        float ou ndarray: Valor(s) transformado(s) pela função sigmoide.
        """
        return 1 / (1 + np.exp(-z))
    
    def fit(self, X, y):
        """
        Treina o modelo de Regressão Logística usando o método de Gradiente Descendente.
        
        Parâmetros:
        X (ndarray): Matriz de características (tamanho: n amostras x m características).
        y (ndarray): Vetor de rótulos (tamanho: n amostras) contendo valores binários (0 ou 1).
        
        O método ajusta os parâmetros (theta) do modelo de modo a minimizar o erro de previsão.
        """
        # Adiciona uma coluna de 1s para o termo de interceptação (bias)
        Xb = np.c_[np.ones((X.shape[0], 1)), X]
        
        # Inicializa os coeficientes theta com zeros
        m, n = Xb.shape
        self.theta = np.zeros(n)

        # Gradiente descendente para otimização dos parâmetros
        for _ in range(self.iterations):
            # Previsões usando a função sigmoide
            predictions = self.sigmoid(Xb.dot(self.theta)) # Fórmula Predictions: sigmoid(X * theta)
            
            # Erros (diferença entre previsões e rótulos reais)
            errors = predictions - y
            
            # Gradiente (derivada do erro com relação aos parâmetros)
            gradient = (1 / m) * Xb.T.dot(errors)  # Fórmula Gradiente: (1/m) * X^T * (erro)
            
            # Atualiza os parâmetros (theta)
            self.theta -= self.learning_rate * gradient # Fórmula Atualização: theta = theta - learning_rate * gradiente

        # O intercepto é o primeiro valor de theta, os demais são os coeficientes
        self.intercept = self.theta[0]
        self.theta = self.theta[1:]

    def predict_proba(self, X):
        """
        Preve a probabilidade de cada amostra pertencer à classe 1 (probabilidade logística).
        
        Parâmetros:
        X (ndarray): Matriz de características (tamanho: n amostras x m características).
        
        Retorna:
        ndarray: Vetor de probabilidades de cada amostra pertencer à classe 1.
        """
        Xb = np.c_[np.ones((X.shape[0], 1)), X]  # Adiciona coluna de 1s para o bias
        return self.sigmoid(Xb.dot(np.r_[self.intercept, self.theta]))  # Calcula a probabilidade

    def predict(self, X, threshold=0.5):
        """
        Preve a classe (0 ou 1) para cada amostra com base em um limiar de probabilidade.
        
        Parâmetros:
        X (ndarray): Matriz de características (tamanho: n amostras x m características).
        threshold (float): Limiar de decisão para determinar a classe. 
                           Se a probabilidade for maior ou igual a este valor, a classe será 1, caso contrário será 0.
        
        Retorna:
        ndarray: Vetor com as classes preditas (0 ou 1) para cada amostra.
        """
        probabilities = self.predict_proba(X)  # Obtém as probabilidades
        return (probabilities >= threshold).astype(int)  # Aplica o limiar e retorna 0 ou 1

File: 2-Regressão Logística/logist_regression.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef double real;

/* =====================================================
   Estruturas básicas
   ===================================================== */

/* Vetor matemático */
typedef struct {
    int size;     // número de elementos
    real *data;   // memória contígua
} Vector;

/* Matriz em layout row-major */
typedef struct {
    int rows;
    int cols;
    real *data;
} Matrix;

/* Macro para acesso matricial
   Evita ponteiros duplos e melhora cache */
#define MAT(m, i, j) m.data[(i) * m.cols + (j)]

/* =====================================================
   Função Sigmoide (estável numericamente)
   =====================================================
   Evita overflow de exp(z) para valores grandes
*/
real sigmoid(real z) {
    if (z >= 0) {
        return 1.0 / (1.0 + exp(-z));
    } else {
        real ez = exp(z);
        return ez / (1.0 + ez);
    }
}

/* =====================================================
   Modelo de Regressão Logística
   ===================================================== */
typedef struct {
    Vector theta;     // coeficientes (SEM bias)
    real intercept;  // termo bias
    real lr;         // learning rate
    int iters;       // iterações
} LogisticReg;

/* =====================================================
   Criação do modelo
   ===================================================== */
LogisticReg logistic_create(real lr, int iters, int n_features) {
    LogisticReg model;
    model.lr = lr;
    model.iters = iters;
    model.intercept = 0.0;

    /* Apenas os coeficientes (bias separado) */
    model.theta.size = n_features;
    model.theta.data = calloc(n_features, sizeof(real));

    return model;
}

/* =====================================================
   Treinamento por Gradient Descent
   ===================================================== */
void logistic_fit(LogisticReg *model, Matrix X, Vector y) {
    int m = X.rows;
    int n = X.cols;

    /* theta completo inclui bias → tamanho n+1 */
    Vector theta;
    theta.size = n + 1;
    theta.data = calloc(n + 1, sizeof(real));

    for (int it = 0; it < model->iters; it++) {

        /* Gradiente separado dos parâmetros */
        Vector grad;
        grad.size = n + 1;
        grad.data = calloc(n + 1, sizeof(real));

        for (int i = 0; i < m; i++) {
            /* z = bias + somatório */
            real z = theta.data[0];
            for (int j = 0; j < n; j++)
                z += MAT(X, i, j) * theta.data[j + 1];

            real p = sigmoid(z);
            real error = p - y.data[i];

            /* Gradiente do bias */
            grad.data[0] += error;

            /* Gradiente dos coeficientes */
            for (int j = 0; j < n; j++)
                grad.data[j + 1] += error * MAT(X, i, j);
        }

        /* Atualização dos parâmetros */
        for (int j = 0; j < theta.size; j++)
            theta.data[j] -= (model->lr / m) * grad.data[j];

        free(grad.data);
    }

    /* Separação final */
    model->intercept = theta.data[0];
    for (int j = 0; j < n; j++)
        model->theta.data[j] = theta.data[j + 1];

    free(theta.data);
}

/* =====================================================
   Probabilidade logística
   ===================================================== */
real logistic_predict_proba(LogisticReg *model, Vector x) {
    real z = model->intercept;
    for (int i = 0; i < x.size; i++)
        z += model->theta.data[i] * x.data[i];
    return sigmoid(z);
}

/* =====================================================
   Classe binária
   ===================================================== */
int logistic_predict(LogisticReg *model, Vector x, real threshold) {
    return logistic_predict_proba(model, x) >= threshold;
}

/* =====================================================
   Main de teste
   ===================================================== */
int main() {
    int m = 4;

    Matrix X = {m, 2, malloc(m * 2 * sizeof(real))};
    Vector y = {m, malloc(m * sizeof(real))};

    real data[4][2] = {{0,0},{0,1},{1,0},{1,1}};
    real labels[4] = {0,0,0,1}; // AND lógico

    for (int i = 0; i < m; i++) {
        MAT(X,i,0) = data[i][0];
        MAT(X,i,1) = data[i][1];
        y.data[i] = labels[i];
    }

    LogisticReg model = logistic_create(0.1, 2000, 2);
    logistic_fit(&model, X, y);

    for (int i = 0; i < m; i++) {
        Vector xi = {2, &MAT(X,i,0)};
        int pred = logistic_predict(&model, xi, 0.5);
        printf("Entrada [%g, %g] -> %d\n",
               MAT(X,i,0), MAT(X,i,1), pred);
    }

    free(X.data);
    free(y.data);
    free(model.theta.data);

    return 0;
}
